### **3. Indexación y selección en arrays con Numpy**

In [2]:
import numpy as np


In [31]:

"""CREACIÓN"""
#Array 1d del 1 al 10
array_1d = np.arange(1, 11)
print(array_1d)

#Array 2D tamaño 3x3
array_2d= np.arange(1,10).reshape(3,3)
print(array_2d)

"""INDEXACIÓN"""
print("Primer elemento 1d:",array_1d[0])
print("Último elemento 1d:",array_1d[-1])
print("Primer elemento primera fila 2D:",array_2d[0,0])
print("Último elemento última fila 2D:",array_2d[-1,-1])
print("Elemento ubicado segunda fila tercera columna 2D:", array_2d[1,2])

"""Selección de filas y columnas"""
print("Slice primera fila completa 2d:", array_2d[0,:])
print("Slice segunda columna completa 2d:", array_2d[:,1])

[ 1  2  3  4  5  6  7  8  9 10]
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Primer elemento 1d: 1
Último elemento 1d: 10
Primer elemento primera fila 2D: 1
Último elemento última fila 2D: 9
Elemento ubicado segunda fila tercera columna 2D: 6
Slice primera fila completa 2d: [1 2 3]
Slice segunda columna completa 2d: [2 5 8]


### **4. Explorando una imagen en formato de Array 3D**

In [9]:
array_3d = np.random.randint(0,256, size=(3, 4, 4))
print(array_3d)

"""INDEXACIÓN"""
print("\nPrimer pixel:", array_3d[0][0])
print("Último pixel:", array_3d[-1][-1])
print("Tercera columna de la segunda fila", array_3d[0][1,2])

"""SLICING"""
print("\nPrimera fila de la imagen", array_3d[:, 0])
print("\Valores del canal azul", array_3d[2])
print("Recorte central de la imagen", array_3d[1, 1:3, 1:3])

[[[244 207  60  48]
  [134 231  11 162]
  [ 40 214 130 180]
  [165  10 199  36]]

 [[152  13 224  18]
  [ 47 100 105 142]
  [239  49  19  47]
  [ 97 253 185 248]]

 [[169 176   6 249]
  [ 10 228 195  14]
  [169 168  33 238]
  [113  27 252  43]]]

Primer pixel: [244 207  60  48]
Último pixel: [113  27 252  43]
Tercera columna de la segunda fila 11

Primera fila de la imagen [[244 207  60  48]
 [152  13 224  18]
 [169 176   6 249]]
\Valores del canal azul [[169 176   6 249]
 [ 10 228 195  14]
 [169 168  33 238]
 [113  27 252  43]]
Recorte central de la imagen [[100 105]
 [ 49  19]]


### **5. Distancia euclidiana con fórmula optimizada**

In [12]:
p1 = np.array([[[[[2]]]], [[[[3]]]], [[[[5]]]], [[[[7]]]], [[[[11]]]]])
p2 = np.array([[[[[1]]]], [[[[4]]]], [[[[6]]]], [[[[8]]]], [[[[10]]]]])
distancia = np.linalg.norm(p1 - p2)
print("Distancia:", distancia)

Distancia: 2.23606797749979


### **6. Análisis del rendimiento académico utilizando normas L1 y L2**

In [ ]:
# Datos de entrada
calificaciones = np.array(
    [
        [88, 78, 64, 92, 57, 70, 88, 68, 72, 60],
        [60, 73, 85, 89, 73, 52, 71, 51, 73, 93],
        [79, 87, 51, 70, 82, 61, 71, 93, 74, 98],
        [76, 91, 77, 65, 64, 96, 93, 52, 86, 56],
        [70, 58, 88, 67, 53, 74, 63, 99, 58, 75],
        [82, 79, 91, 88, 66, 78, 79, 63, 59, 89],
        [56, 66, 73, 77, 75, 80, 70, 64, 88, 60],
        [80, 62, 79, 85, 92, 72, 80, 69, 78, 66],
        [88, 90, 69, 77, 66, 72, 65, 73, 92, 81],
        [60, 80, 69, 94, 81, 88, 93, 60, 58, 79],
        [78, 72, 74, 81, 67, 93, 60, 72, 85, 92],
        [86, 90, 70, 64, 93, 58, 61, 95, 79, 84],
        [85, 77, 91, 85, 61, 63, 58, 68, 65, 90],
        [71, 79, 69, 58, 75, 78, 62, 82, 77, 59],
        [87, 76, 72, 68, 88, 65, 75, 90, 69, 79],
        [78, 71, 94, 67, 72, 88, 58, 79, 91, 67],
        [67, 64, 85, 81, 70, 84, 79, 73, 62, 77],
        [90, 61, 60, 91, 72, 89, 77, 75, 76, 83],
        [94, 83, 80, 59, 65, 75, 63, 92, 68, 60],
        [80, 81, 68, 95, 72, 93, 89, 59, 81, 85],
        [88, 63, 90, 77, 69, 72, 85, 78, 61, 93],
        [91, 65, 79, 92, 63, 74, 87, 70, 92, 85],
        [92, 80, 76, 72, 81, 87, 64, 74, 70, 67],
        [72, 83, 91, 58, 67, 88, 81, 76, 74, 92],
        [84, 91, 81, 77, 80, 63, 70, 79, 89, 65],
        [66, 65, 77, 58, 79, 61, 86, 72, 64, 74],
        [94, 74, 83, 89, 77, 68, 81, 92, 70, 68],
        [73, 72, 82, 80, 79, 64, 69, 60, 76, 79],
        [82, 79, 75, 89, 85, 81, 62, 64, 60, 92],
        [87, 88, 80, 74, 85, 90, 75, 68, 68, 71],
        [81, 79, 85, 76, 64, 85, 71, 77, 67, 93],
        [82, 89, 77, 72, 66, 84, 92, 60, 67, 65],
        [90, 63, 94, 67, 88, 93, 79, 64, 88, 80],
        [81, 90, 72, 76, 74, 93, 88, 84, 80, 65],
        [86, 75, 67, 81, 93, 79, 91, 64, 78, 74],
        [69, 85, 90, 91, 60, 76, 84, 78, 74, 64],
        [67, 91, 94, 78, 81, 73, 62, 67, 60, 92],
        [85, 79, 73, 60, 85, 90, 82, 75, 74, 68],
        [91, 83, 77, 92, 74, 68, 64, 92, 90, 92],
        [65, 93, 88, 82, 92, 67, 79, 68, 81, 93],
        [92, 79, 90, 92, 72, 80, 61, 75, 86, 90],
        [80, 67, 76, 84, 92, 90, 79, 93, 74, 88],
        [60, 79, 90, 93, 64, 61, 85, 76, 92, 72],
        [88, 67, 72, 64, 80, 90, 79, 75, 85, 81],
        [79, 92, 75, 72, 92, 81, 79, 70, 68, 68],
        [65, 60, 79, 89, 79, 82, 86, 64, 60, 81],
        [92, 81, 60, 79, 94, 78, 80, 91, 85, 68],
        [79, 64, 81, 72, 68, 75, 80, 94, 79, 61],
        [79, 94, 64, 89, 80, 85, 74, 82, 65, 91],
    ]
)


def calificaciones_l2(datos):
    """
    Normaliza cada fila de la matriz usando la norma L2.

    Parámetros:
    datos (ndarray): Matriz con los valores a normalizar.

    Retorna:
    ndarray: Matriz con los valores normalizados.
    """
    return datos / np.linalg.norm(datos, axis=1, keepdims=True)
    # Cuando usamos axis=1 en una función de NumPy, estamos indicando que la operación se realizará a lo largo de las filas, es decir, afecta a las columnas


def asignaturas_l2(datos):
    """
    Normaliza cada fila de la matriz usando la norma L2.

    Parámetros:
    datos (ndarray): Matriz con los valores a normalizar.

    Retorna:
    ndarray: Matriz con los valores normalizados.
    """
    return datos / np.linalg.norm(datos, axis=0, keepdims=True)


# Normalización de las calificaciones
calificaciones_normalizadas = calificaciones_l2(calificaciones)
sumatorio = np.sum(calificaciones_normalizadas, axis=1)
print("Sumatorio calificaciones:", sumatorio)
""" for i, calificacion in enumerate(calificaciones_normalizadas):
    print(f"Alumno {i+1}: {calificacion}") """


# Normalización de las asignaturas
asignaturas_normalizadas = asignaturas_l2(calificaciones)

sumatorio_filas = np.sum(asignaturas_normalizadas, axis=0)
print("Sumatorio asignaturas:", sumatorio_filas)


Sumatorio calificaciones: [3.12307762 3.10621962 3.1135639  3.10169671 3.10536531 3.13275609
 3.13597111 3.1423133  3.13872561 3.11719673 3.13599435 3.11880261
 3.12132529 3.14103949 3.14338727 3.12954772 3.14480191 3.13316674
 3.119699   3.13444775 3.13254857 3.13450148 3.14348648 3.13482883
 3.14166543 3.13981624 3.14097437 3.14838834 3.13158412 3.14581251
 3.14405674 3.132219   3.13100031 3.14549032 3.14201171 3.13572714
 3.12369819 3.1431708  3.13795456 3.13635672 3.14010203 3.14693417
 3.12430286 3.14507978 3.14378806 3.13133026 3.1376334  3.13943549
 3.1393785 ]
Sumatorio asignaturas: [6.94456999 6.94026697 6.94482158 6.93168944 6.93326078 6.93189681
 6.93672459 6.91546516 6.93526259 6.92004648]


- **¿Por qué la norma L2 nos da una idea de la magnitud total de los valores, penalizando más las notas altas?**
La norma L2 mide la magnitud total de un conjunto de valores sumando sus cuadrados y sacando la raíz cuadrada. Penaliza más los valores grandes porque al elevarlos al cuadrado, su impacto crece mucho más rápido que los valores pequeños.

- **¿Cómo afecta keepdims=True en las operaciones realizadas?**
Asegura que la respuesta conserve las dimensiones originales en lugar de reducirlas.

### **7. Producto cruzado para calcular variaciones en la ruta de un avión**

In [10]:
vuelo_previsto = [3, 4, 0]
vuelo_real = [4, 3, 1]

producto_cruzado = np.cross(vuelo_previsto, vuelo_real)
print("Producto cruzado:", producto_cruzado)


Producto cruzado: [ 4 -3 -7]


**¿En qué dirección ocurre la desviación?**
El componente del eje x (4) indica que hay una desviación relativamente leve a la derecha en ese eje. Para el eje y, hay una desviación significativa (-3) al igual que en el eje Z donde la desviación es muy relevante (-7).

**¿Cómo se podría corregir la trayectoria del avión basándonos en este cálculo?**
El avión debe moverse en la dirección hacia la izquierda en el eje 𝑥, hacia arriba en el eje Y además de hacia arriba en el eje Z

### **8. Análisis del rendimiento de un atleta con producto escalar**

In [21]:
tiempos = np.array([11, 22, 50, 110, 250])
importancia = np.array([0.4, 0.3, 0.2, 0.1, 0.05])

producto_escalar = np.dot(tiempos, importancia)
print("Producto escalar", producto_escalar)

Producto escalar 44.5


**Explica qué significa el resultado obtenido en términos del rendimiento del atleta**
El resultado obtenido (44.5) representa un índice de rendimiento del atleta que combina sus tiempos en cada prueba con la importancia asignada a cada una. Un valor más bajo del índice indica un mejor rendimiento.

**¿Cómo afectaría al resultado si el atleta mejora su tiempo en la prueba de 100m?**
Como la prueba de 100m tiene la mayor importancia (0.4), mejorar su tiempo reduciría significativamente el índice de rendimiento, reflejando una mejora general en su desempeño.

**¿Qué pasaría si le damos más importancia a las pruebas de resistencia?**
El resultado dependería más de los tiempos en pruebas largas (800m y 1500m), por lo que si el atleta tiene tiempos altos en estas distancias, su índice de rendimiento aumentaría.

**¿Cómo se podría utilizar esta métrica para comparar a diferentes atletas?**
Se puede comparar fácilmente a los atletas usando el índice calculado: un valor menor indica un mejor rendimiento por lo que sería sencillo clasificarlos.

### **9. Normalización de datos con la norma L2 en rendimiento deportivo**

In [19]:
datos_jugadores = np.array([
    [10, 7, 300, 25, 40],  
    [15, 10, 250, 30, 50], 
    [5, 12, 280, 20, 35],   
    [8, 6, 310, 15, 38],   
    [12, 9, 270, 28, 45]   
])

def normalizacion_l2(datos):
    """
    Normaliza cada fila de la matriz usando la norma L2.

    Parámetros:
    datos (ndarray): Matriz con los valores a normalizar.

    Retorna:
    ndarray: Matriz con los valores normalizados.
    """
    return datos / np.linalg.norm(datos, axis=1, keepdims=True)
datos_normalizados = datos_jugadores / np.linalg.norm(datos_jugadores, axis = 1, keepdims=True)
for i, dato in enumerate(datos_normalizados):
    print(f"Jugador {i+1}: Goles:{dato[0]} Asistencias: {dato[1]} Pases = {dato[2]} Regates = {dato[3]}\
Tiros a Puerta = {dato[4]}")
    
rendimientos = np.sum(datos_normalizados, axis=1)
rendimientos_ordenados = np.sort(rendimientos)[::-1]
print(f"1º: {rendimientos_ordenados[0]} 2º: {rendimientos_ordenados[1]} 3º: {rendimientos_ordenados[2]} 4º:\
 {rendimientos_ordenados[3]} 5º: {rendimientos_ordenados[4]}")




Jugador 1: Goles:0.03290221417307103 Asistencias: 0.02303154992114972 Pases = 0.9870664251921308 Regates = 0.08225553543267757Tiros a Puerta = 0.13160885669228412
Jugador 2: Goles:0.058288150543410064 Asistencias: 0.03885876702894004 Pases = 0.971469175723501 Regates = 0.11657630108682013Tiros a Puerta = 0.19429383514470022
Jugador 3: Goles:0.01765627426015132 Asistencias: 0.04237505822436317 Pases = 0.988751358568474 Regates = 0.07062509704060528Tiros a Puerta = 0.12359391982105924
Jugador 4: Goles:0.025572159912226913 Asistencias: 0.019179119934170187 Pases = 0.990921196598793 Regates = 0.047947799835425464Tiros a Puerta = 0.12146775958307784
Jugador 5: Goles:0.043547487891216305 Asistencias: 0.032660615918412225 Pases = 0.9798184775523668 Regates = 0.1016108050795047Tiros a Puerta = 0.16330307959206114
1º: 1.3794862295273713 2º: 1.320940466033561 3º: 1.2568645814113133 4º: 1.243001707914653 5º: 1.2050880358636933


**¿Cómo cambian los valores después de la normalización?**
Se convierten en valores proporcionales del 0 al 1, sus cuadrados de cada fila es 1.

**¿Por qué la normalización L2 es útil cuando los datos tienen escalas diferentes?**
Porque ajusta todos los valores a una escala común, evitando que atributos con valores grandes (como pases) dominen sobre otros más pequeños (como goles).

**¿Qué pasaría si normalizamos por **columnas (axis=0)** en lugar de por filas?**
Si normalizamos por columnas (axis=0), podríamos comparar a los jugadores en cada atributo específico, pero perderíamos la visión completa de su rendimiento individual.

### **10. Hablando de pilotos de carreras**

In [29]:
rendimiento_pilotos = np.array([
    [3, 2, 5, 5, 5],  
    [9, 10, 2, 9, 5], 
    [5, 6, 7, 6, 3],  
    [9, 7, 4, 10, 2],  
    [10, 10, 6, 8, 4],  
    [6, 1, 10, 9, 3],  
    [9, 8, 1, 2, 3],  
    [3, 4, 6, 1, 8],  
    [2, 2, 1, 5, 3], 
    [5, 9, 4, 9, 3]   
])

#Transformación de arrays:
rendimientos_reorganizados = rendimiento_pilotos.reshape(5,10)
print("Rendimientos reorganizados:\n",rendimientos_reorganizados)

rendimientos_1d = rendimiento_pilotos.ravel()
print("Conversión a 1d:\n", rendimientos_1d)

#Operaciones aritméticas
media_pilotos = np.sum(rendimiento_pilotos, axis=1) / len(rendimiento_pilotos[0])
media_carrera = np.sum(rendimiento_pilotos, axis=0) / len(rendimiento_pilotos)

print("Promedio de posiciones por piloto:",media_pilotos)
print("Promedio de posiciones por carrera :",media_carrera)

penalizacion = 2
rendimiento_penalizado = rendimiento_pilotos + penalizacion

media_pilotos_penalizado = np.sum(rendimiento_penalizado, axis=1) / len(rendimiento_penalizado[0])
media_carrera_penalizado = np.sum(rendimiento_penalizado, axis=0) / len(rendimiento_penalizado)

print("Promedio de posiciones por piloto (penalizado):", media_pilotos_penalizado)
print("Promedio de posiciones por carrera (penalizado):", media_carrera_penalizado)

#Producto escalar
piloto_1 = rendimiento_pilotos[0]
piloto_2 = rendimiento_pilotos[1]
producto_escalar = np.dot(piloto_1, piloto_2)
print("Producto escalar entre piloto 1 y piloto 2:", producto_escalar)

#Producto cruzado
piloto_1_3d = rendimiento_pilotos[0][:3]
piloto_2_3d = rendimiento_pilotos[1][:3]
producto_cruzado = np.cross(piloto_1_3d, piloto_2_3d)
print("Producto cruzado entre piloto 1 y piloto 2 (primeras 3 carreras):", producto_cruzado)
print("El producto cruzado", producto_cruzado, "indica que en la primera carrera el piloto 1 tuvo peor rendimiento que el 2 a diferencia de las \
otras dos carreras, donde el piloto 1 tuvo mejor rendimiento, especialmente en la segunda.")

#Extra: Normalización usando la norma L2
rendimiento_normalizado = rendimiento_pilotos / np.linalg.norm(rendimiento_pilotos, axis=1, keepdims=True)
print("Rendimiento normalizado:\n", rendimiento_normalizado)


Rendimientos reorganizados:
 [[ 3  2  5  5  5  9 10  2  9  5]
 [ 5  6  7  6  3  9  7  4 10  2]
 [10 10  6  8  4  6  1 10  9  3]
 [ 9  8  1  2  3  3  4  6  1  8]
 [ 2  2  1  5  3  5  9  4  9  3]]
Conversión a 1d:
 [ 3  2  5  5  5  9 10  2  9  5  5  6  7  6  3  9  7  4 10  2 10 10  6  8
  4  6  1 10  9  3  9  8  1  2  3  3  4  6  1  8  2  2  1  5  3  5  9  4
  9  3]
Promedio de posiciones por piloto: [4.  7.  5.4 6.4 7.6 5.8 4.6 4.4 2.6 6. ]
Promedio de posiciones por carrera : [6.1 5.9 4.6 6.4 3.9]
Promedio de posiciones por piloto (penalizado): [6.  9.  7.4 8.4 9.6 7.8 6.6 6.4 4.6 8. ]
Promedio de posiciones por carrera (penalizado): [8.1 7.9 6.6 8.4 5.9]
Producto escalar entre piloto 1 y piloto 2: 127
Producto cruzado entre piloto 1 y piloto 2 (primeras 3 carreras): [-46  39  12]
El producto cruzado [-46  39  12] indica que en la primera carrera el piloto 1 tuvo peor rendimiento que el 2 a diferencia de las otras dos carreras, donde el piloto 1 tuvo mejor rendimiento, especialmente en

### **11. Ejercicio de investigación de Numpy**


- **¿Qué es NumPy y para qué se utiliza en programación científica?** <br>
NumPy, cuyo nombre proviene de "Numerical Python", es una biblioteca fundamental para realizar cálculos numéricos en Python. Facilita la manipulación de arrays multidimensionales y proporciona herramientas para llevar a cabo operaciones matemáticas complejas de manera eficiente, lo que la hace ideal para análisis de datos y aplicaciones científicas.

- **Historia de NumPy** <br>
La historia de NumPy se inicia en 1995 con la creación de Numeric por Jim Hugunin, quien diseñó la primera biblioteca de Python para cálculos numéricos. Posteriormente, en 2005, Travis Oliphant integró las características de Numeric y Numarray, dando origen a NumPy, una solución más optimizada y consistente. NumPy introdujo los arrays multidimensionales (ndarray), acelerando las operaciones mediante código en C y convirtiéndose en la base para bibliotecas como SciPy, Pandas y TensorFlow.

- **¿Cuál es la diferencia entre una lista de Python y un array de NumPy?** <br>
Las listas en Python pueden contener elementos de distintos tipos y cambian de tamaño dinámicamente, mientras que los arrays de NumPy son estructuras homogéneas por lo que todos sus elementos deben ser del mismo tipo de dato. Los arrays de NumPy están diseñados para realizar cálculos numéricos de manera mucho más rápida y eficiente en comparación con las listas estándar de Python.

- **¿Cómo se crea un array en NumPy? Explicar y dar ejemplos de:** <br>
     Se pueden crear arrays en NumPy utilizando la función **numpy.array()**. <br>
     - **Un array unidimensional.** <br>
     `array = np.array([1, 2, 3])` 
     - **Un array bidimensional.** <br>
     `array = np.array([[1, 2, 3], [4, 5, 6]])`<br>
     - **Un array de valores inicializados con ceros, unos y valores aleatorios.**<br>
     `ceros = np.zeros((2, 4))`<br>
     `unos = np.ones((2, 4))`<br>
     `random = np.random.rand(2, 4)`<br>

- **¿Cómo acceder a los elementos de un array y cómo modificar sus valores?** <br>
Los elementos de un array se pueden acceder y cambiar utilizando índices o slicing. <br>
     `valor = array[0]`   **Primer elemento**
     `valor_final = array[-1]`  **Último elemento**
     `valor_matriz = array[0, 1]` **Primera fila, segunda columna**

- **¿Qué son los ejes (axis) en un array multidimensional? Explica con ejemplos prácticos.** <br>
Los ejes en un array multidimensional indican las dimensiones en las que se pueden realizar operaciones. En un array de una dimensión, solo hay un eje (axis 0), mientras que en un array de dos dimensiones existen dos ejes:

**axis=0** se refiere a las filas.<br>
**axis=1** se refiere a las columnas.

**Ejemplo:**<br>
`array = np.array([[1, 2, 3], [4, 5, 6]])`<br>
`Suma con axis=0 (columnas):", np.sum(array, axis=0))` **Daría [5,7,9] como resultado** <br>
`Suma con axis=1 (filas):", np.sum(array, axis=1))`   **Daría [6,15] como resultado**   <br>

- **Explica la diferencia entre np.dot() (producto escalar) y np.cross() (producto cruzado).** <br>
El producto escalar, calculado con **np.dot()**, proporciona un solo valor numérico que mide la similitud entre dos vectores multiplicando sus componentes y sumándolos.<br>
El producto cruzado, calculado con **np.cross()**, genera un nuevo vector perpendicular a los dos vectores originales, útil en aplicaciones geométricas y de física para determinar direcciones ortogonales en el espacio tridimensional.

- **¿Cómo se pueden transformar las dimensiones de un array? Explica con ejemplos las funciones reshape() y flatten().** <br>
La función **reshape()** permite cambiar las dimensiones de un array, siempre que se mantenga el número total de elementos.<br>

**Ejemplo:**<br>
 `array = np.array([1, 2, 3, 4, 5, 6])`<br>
 `array_2d = array.reshape(2, 3)`<br>
 **array_2d nos daría como resultado: [[1 2 3] [4 5 6]]** <br>

La función **flatenn()** convierte un array multidimensional en un array unidimensional. <br>
**Ejemplo** <br>
`array2_d.flatten()` <br>
 **array2_d pasaría a tener el valor: [1 2 3 4 5 6]** <br>

- **¿Qué es Matplotlib y en qué situaciones se usa en programación?** <br>
Matplotlib es una biblioteca de visualización de datos en Python que permite crear gráficos en 2D y 3D, como gráficos de líneas, barras y dispersión. Es utilizado tanto en análisis de datos como generación de informes o aplicaciones científicas para representar datos de manera comprensible y visualmente atractiva.

# 12. Transformación de datos sesgados con raíz cuadrada
    


In [3]:
    
""" 1. Crea un array de valores que sigan una progresión exponencial, por ejemplo: `[1, 10, 100, 1000, 10000]`.
    2. Aplica la transformación de raíz cuadrada utilizando `np.sqrt()`.
    3. Imprime los valores originales y los valores transformados.
"""
exponenciales = np.array([1,10,100,1000,10000])
raiz = np.sqrt(exponenciales)

print(f"Valores originales:{exponenciales}, valores transformados: {raiz}")

Valores originales:[    1    10   100  1000 10000], valores transformados: [  1.           3.16227766  10.          31.6227766  100.        ]


# 13. Cálculo del RMSE (Error Cuadrático Medio)

In [ ]:
"""
1. Define dos arrays: uno con valores reales (`y`) y otro con valores predichos (`y_pred`).
2. Calcula el error cuadrático medio (MSE) elevando al cuadrado las diferencias entre `y` y `y_pred`, promediando los valores y tomando la raíz cuadrada.
3. Imprime el resultado del RMSE.
4. Repite el cálculo utilizando la función `root_mean_squared_error` de `sklearn.metrics`.
"""

y = np.array([4, 6, 7, 2, 5])
y_pred = np.array([1.5, 3.2, 5.4, 7.2, 2.1])

error_cuadratico = (y - y_pred) ** 2
mse = np.mean(error_cuadratico)
rmse = np.sqrt(mse)

print(f"RMSE: {rmse:.2f}")

from sklearn.metrics import root_mean_squared_error 
rmse_scikit = root_mean_squared_error(y, y_pred)
print(f"RMSE con la función root_mean_squared_error: {rmse_scikit:.2f}")


# 14. Cálculo de probabilidades con Softmax

In [ ]:
"""
1. Define un array con tres valores representando logits de un modelo.
2. Aplica la función exponencial a cada valor y normaliza dividiendo entre la suma total.
3. Imprime las probabilidades resultantes.
4. Implementa el cálculo de softmax utilizando las funciones de `scipy.special.softmax` y `sklearn.utils.extmath.softmax`. Compara los resultados.
"""
logits = np.array([1.4, 2.1, 0.8])
valores_exp = np.exp(logits)
probabilidades = valores_exp / np.sum(valores_exp)

print("Probabilidades:", probabilidades)

#Función softmax de Scikit-learn
from sklearn.utils.extmath import softmax  
probabilidades_sklearn = softmax(logits.reshape(1, -1))
print("Probabilidades softmax con Scikit-learn:", probabilidades_sklearn)

#Función softmax de scipy
from scipy.special import softmax  
probabilidades_scipy = softmax(logits)
print("Probabilidades softmax con Scipy:", probabilidades_scipy)



